router/h2/src/main/scala/io/buoyant/router/h2/StreamStatsFilter.scala

```diff
@@ -41,46 +44,118 @@ class StreamStatsFilter(statsReceiver: StatsReceiver)
+              // callback we were placing on `stream.onEnd` gets
+              // clobbered – see above comment
+              frameBytes.add(streamFrameBytes)
+              classifyFrame(Return(Some(frame)))
+              streamStatsT(Return(Some(frame)))
+            }
+          case Throw(e) =>
+            classifyFrame(Throw(e))
+            streamStatsT(Throw(e))
+        }
+      stream
+    }
+
+  }
+
+  // total number of requests received
   private[this] val reqCount = statsReceiver.counter("requests")
 
   private[this] val reqStreamStats =
-    new StreamStats(statsReceiver.scope("request", "stream"))
+    new FrameStats(statsReceiver.scope("request", "stream"))
   private[this] val rspStreamStats =
-    new StreamStats(statsReceiver.scope("response", "stream"))
+    new FrameStats(statsReceiver.scope("response", "stream"))
   private[this] val totalStreamStats =
     new StreamStats(
       statsReceiver.scope("stream"),
       durationName = Some("total_latency")
     )
-  private[this] val rspFutureStats =
-    new StreamStats(
-      statsReceiver,
-      durationName = Some("request_latency"),
-      successName = Some("")
-    )
 
-  override def apply(req: Request, service: Service[Request, Response]): Future[Response] = {
+  private[this] val reqLatency = statsReceiver.stat("request_latency_ms")
+
+  // overall successes stat from response classifier
+  private[this] val successes = statsReceiver.counter("success")
+  // overall failures stat from response classifier
+  private[this] val failures = statsReceiver.counter("failures")
+
+  override def apply(req0: Request, service: Service[Request, Response]): Future[Response] = {
     reqCount.incr()
     val reqT = Stopwatch.start()
+    val req1 = Request(req0.headers, reqStreamStats(reqT)(req0.stream))
 
-    req.stream.onEnd.respond(reqStreamStats(reqT))
+    @inline def classify(rsp: Try[Response])(frame: Try[Option[Frame]]): Unit =
+      classifier(H2ReqRep(req1, rsp.map((_, frame)))) match {
+        case Successful(_) => successes.incr()
+        case _ => failures.incr()
+      }
 
-    service(req)
+    service(req1)
+      .transform {
+        case Return(response) =>
+          val rspT = Stopwatch.start()
+          val stream = rspStreamStats(rspT, classify(Return(response)))(response.stream)
+          Future.value(Response(response.headers, stream))
+        case Throw(e) =>
+          classify(Throw(e))(Throw(e))
+          Future.exception(e)
+      }
       .respond { result =>
-        rspFutureStats(reqT)(result)
+        reqLatency.add(reqT().inMillis)
         val stream = result match {
-          case Return(rsp) => req.stream.onEnd.join(rsp.stream.onEnd)
-          case Throw(_) => req.stream.onEnd
+          case Return(rsp) =>
+            // this is the call i mentioned above that seems to be wiping out
+            // the `.onEnd.respond { }` callback in `frameStats.apply()`...
```

Comment:
that would surprise me.  if there's something improper happening here we should get a repro of it.